src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -355,19 +357,30 @@ protected GitTool resolveGitTool(String gitTool) {
     @Override
     protected SCMRevision retrieve(@NonNull final SCMHead head, @NonNull TaskListener listener)
             throws IOException, InterruptedException {
+        final GitSCMSourceContext context = new GitSCMSourceContext<>(null, SCMHeadObserver.none()).withTraits(getTraits());
         return doRetrieve(new Retriever<SCMRevision>() {
                               @Override
                               public SCMRevision run(GitClient client, String remoteName) throws IOException, InterruptedException {
-                                  for (Branch b : client.getRemoteBranches()) {
-                                      String branchName = StringUtils.removeStart(b.getName(), remoteName + "/");
-                                      if (branchName.equals(head.getName())) {
-                                          return new SCMRevisionImpl(head, b.getSHA1String());
+                                  if (head instanceof GitTagSCMHead) {
+                                      try {
+                                          ObjectId objectId = client.revParse(Constants.R_TAGS + head.getName());
+                                          return new GitTagSCMRevision((GitTagSCMHead) head, objectId.name());
+                                      } catch (GitException e) {
+                                          // tag does not exist
+                                          return null;
+                                      }
+                                  } else {
+                                      for (Branch b : client.getRemoteBranches()) {
+                                          String branchName = StringUtils.removeStart(b.getName(), remoteName + "/");
+                                          if (branchName.equals(head.getName())) {
+                                              return new SCMRevisionImpl(head, b.getSHA1String());
+                                          }
                                       }
                                   }
                                   return null;
                               }
                           },
-                new GitSCMSourceContext<>(null, SCMHeadObserver.none()).withTraits(getTraits()),
+                context,
```

Comment:
Why was this moved up? AFAICT you are not using this variable until now.

Reply:
May have been a left-over of some earlier refactoring where I was filtering the revision discovery based on the context, e.g. I may have had

```
if (context.wantTags() && head instanceof GitTagSCMHead) {
...
} else if (context.wantBranches()) {
...
}
```

within the closure... but then I decided that conflicted. If you have a valid `SCMHead`, even if it would not be discovered by the `fetch(TaskListener)` we should still report its revision (as long as it exists)